core/src/main/scala/kafka/zk/KafkaZkClient.scala

```diff
@@ -893,14 +896,29 @@ class KafkaZkClient private[zk] (zooKeeperClient: ZooKeeperClient,
    * @return list of child node names
    */
   def getChildren(path : String): Seq[String] = {
-    val getChildrenResponse = retryRequestUntilConnected(GetChildrenRequest(path, registerWatch = true))
+    val getChildrenResponse = retryRequestUntilConnected(
+      if (shouldPaginate(path)) {
+        debug(s"upgrading GetChildrenRequest to GetChildrenPaginatedRequest for '${path}'")
+        GetChildrenPaginatedRequest(path, registerWatch = true)
+      } else
+        GetChildrenRequest(path, registerWatch = true))
     getChildrenResponse.resultCode match {
       case Code.OK => getChildrenResponse.children
       case Code.NONODE => Seq.empty
       case _ => throw getChildrenResponse.resultException.get
     }
   }
 
+  /**
+   * Returns true if ZK pagination is enabled and the specified path may have an excessive number of children.
+   * (Sole concerns currently are /brokers/topics and /config/topics, both of which are nominally unbounded.)
+   * @param path Zookeeper path to enumerate
+   * @return whether to use GetChildrenPaginated for the request due to possible excessive response size
+   */
+  private def shouldPaginate(path : String): Boolean = {
+    paginateTopics && path.endsWith("/topics")
```

Comment:
Can we make this more explicitly as "path.endsWith(/brokers/topics) ||  path.endsWith(/brokers/topics)" such that it is clear on which Znodes we would use pagenation and prevents this being mis-applied to other possible nodes that also end with "/topics"? And we could also have this list of nodes using pagenation to be configurable.

Reply:
That's possible, but what if there's another full topics znode hidden somewhere in the tree? (Which already happened.) Or what if it happens in the future? More realistically, Unix paths tree "foo/bar" and "foo//bar" and "foo/./bar" as the same; it seemed prudent to assume ZK might as well (at least the first two), or else a user request might kill a processing thread, which could lead to a kind of DoS attack.

Reply:
Pros for listing explicitly as I suggested: we do not use pagenation for un-wanted znodes (e.g., someParentNode/topics); and we can make it configurable, such that if there is unfounded znode that you need pagination, we can just update the config. And the hidden znode that need pagination may not even end with '/topics'. So we can have a list of strings in the configs.

Reply:
That's a distinctly bigger change than this one, however (not "big" overall, just more things to check and test), and we're pressed for time on the release. Having tested pagination on both single-page and multi-page znodes at this point, I don't see that it matters greatly if it's used for random other znodes in a one-off sense; other than the lack of "stat" support and the use of an explicit thread pool, it seems to behave exactly as the normal getChildren() call.

Configurability isn't a terrible idea -- though there's always a cost associated with more configuration -- but I'm not totally convinced it's needed, and in any case I don't want to hold up this release adding it.

Reply:
Deferring the configurability possibility per offline discussion.